File: optimized/fib-mem.go
package main

func fib(n int, cache *[]uint64) uint64 {
	if (*cache)[n] == 0 {
		a := fib(n-2, cache)
		b := fib(n-1, cache)
		(*cache)[n] = a+b
	}
	return (*cache)[n]
}

func main() {
	const n = 46
	cache := make([]uint64, n+1, n+1)
	cache[0] = 1
	cache[1] = 1
	println(fib(n, &cache))
}

File: optimized/fib_mem.nim
func fib(n: uint64, cache: var openArray[uint64]): uint64 =
  if n <= 1:
    return 1
  var
    a = cache[n.int - 1]
    b = cache[n.int - 2]
  if a == 0:
    a = fib(n - 1, cache)
    cache[n.int - 1] = a
  if b == 0:
    b = fib(n - 2, cache)
    cache[n.int - 2] = b
  return a + b

var cache: array[46, uint64]
echo fib(46, cache)
